### ISSUE TITLE
API response may include more unique visitors than actions

### ISSUE THREAD

**Issue by mattab** [matomo-org/matomo#9357]
It has been reported at least once that an API request was returning more "Unique visitors" than "Actions". This is un-expected as we should always have: `nb_uniq_visitors <= nb_visits <= nb_actions`

Here is the API response that was at least once:

```
[nb_uniq_visitors] => 2
[nb_users] => 0
[nb_visits] => 1
[nb_actions] => 1
[nb_visits_converted] => 0
[bounce_count] => 1
[sum_visit_length] => 0
[max_actions] => 1
[bounce_rate] => 100%
[nb_actions_per_visit] => 1
[avg_time_on_site] => 0
```

(request was `method=VisitsSummary.get&idSite=10435&period=range&date=2015-11-16,2015-12-03` )

Questions:
- How can we reproduce this issue?
- What is the root cause?

Only one possible explanation we came up with so far:
- imagine the `period=range` request includes today's data, and if the "Unique visitors processing" archiving query runs a few seconds after the standard metrics processing... then imagine that a new visitor visits today's in the few seconds right after the standard metrics processing and right before the unique visitors processing... then we could get nb_uniq_visitors = 2 while nb_visits = 1...

**Comment by mattab** [comment]
I still cannot reproduce the issue but it was reported again:

```
 [nb_uniq_visitors] => 11 [nb_users] => 0 [nb_visits] => 9 [nb_actions] => 18 [nb_visits_converted] => 0 [bounce_count] => 4 [sum_visit_length] => 4367 [max_actions] => 4 [bounce_rate] => 44% [nb_actions_per_visit] => 2 [avg_time_on_site] => 485
```

One solution for this likely edge-case issue would be to do: `nb_uniq_visitors = min(nb_uniq_visitors, nb_visits)`.

**Comment by quba** [comment]
@tsteur, @mattab,

personally, I don't like this approach to resolve this issue. This may work as a quick fix, but for the long term it looks like a hack. 

Here are my ideas:
1) (preferred but risky) After starting the archiving process, save current timestamp and calculate *ALL* reports for given site ID based on this timestamp. This will make sure that all reports are consistent and based on the same data set. Of course there will be some issue with the `visit_last_action_time` (we may report less visits for today) but still this is better than random differences because of new visits during archiving process. 
2) Calculate the number of uniques at the beginning of archive process for each period.

**Comment by tsteur** [comment]
This might be something for the archiver refactoring https://github.com/piwik/piwik/issues/7470 where we can think about such cases. Right now it's not really doable since we need to rewrite the archiver first (and possibly also allow different backends like elastic search etc)

**Comment by jczyzniejewski-epi** [comment]
```
Matomo version: 5.1.1
MySQL version: 11.5.2-MariaDB
PHP version: 8.2.23
```


I have observed a situation that breaks the constraint mentioned in this ticket:

`nb_uniq_visitors <= nb_visits <= nb_actions`

My request:

`/?module=API&method=Actions.getPageUrl&pageUrl=/home&idSite=1&period=day&date=today&format=JSON&token_auth=<token>`

Matomo response:

```
[
  {
    "label": "/home",
    "nb_visits": 4,
    "nb_uniq_visitors": 7,
    "nb_hits": 15,
    "sum_time_spent": 1822,
    "entry_nb_uniq_visitors": 3,
    "entry_nb_visits": "3",
    "entry_nb_actions": "20",
    "entry_sum_visit_length": "749",
    "entry_bounce_count": "0",
    "exit_nb_uniq_visitors": 1,
    "exit_nb_visits": "1",
    "avg_time_on_page": 121,
    "bounce_rate": "0%",
    "exit_rate": "25%",
    "url": "https://<domain>/home"
  }
]
```

Is the inequality `nb_uniq_visitors <= nb_visits` always true, or is there an explanation for such a response (`nb_uniq_visitors=7`, `nb_visits=4`)?

Reproduction scenario: visit a page multiple times using a browser in incognito/private mode (I use Firefox 132.0.2).

Also: I use `period=day` because it is the only one that returns `nb_uniq_visitors`. Is it possible to enable this field for other values of `period` parameter? I have configured Matomo server according to the instructions from https://matomo.org/faq/how-to/faq_113/ but `nb_uniq_visitors` is still missing for `period=month`. Should I open a new ticket?

**Comment by sgiehl** [comment]
@jczyzniejewski-epi Ist the tracked site by any chance use (cookie) consent or the user id feature?

**Comment by jczyzniejewski-epi** [comment]
@sgiehl No. I only use default Matomo tracking code + Custom Dimensions + Event Tracking.
